### PR TITLE
[Refactor][Distributed] Remove all_to_all patch and switch lm-head TP to all_to_all_single

### PR DESCRIPTION
**Pull request by xuchi-0808**
### What this PR does / why we need it?

Removes the `GroupCoordinator.all_to_all` monkey-patch from vllm-ascend and switches all remaining consumers to a new `lmhead_all_to_all` helper that uses `torch.distributed.all_to_all_single`, returning to the upstream PyTorch standard interface.

This is a re-opening of #11086.

**Background** — the old path required two layers of patch:
```
consumer → get_lmhead_tp_group().all_to_all()   ← GroupCoordinatorPatch.all_to_all
         → device_communicator.all_to_all()      ← NPUCommunicator.all_to_all
         → dist.all_to_all(list version)          ← real HCCL call
```

**Changes:**
- Delete `GroupCoordinatorPatch.all_to_all` and `NPUCommunicator.all_to_all` (preserving `_NpuAll2AllManager`).
- Add `lmhead_all_to_all(logits, comm_group)` helper in `vocab_parallel_embedding.py`, used by all 4 consumers:
  - `vocab_parallel_embedding.py:_get_logits_lmheadtp`
  - `spec_decode/llm_base_proposer.py:_run_merged_draft` (2 sites, reduce_sample branches)
  - `spec_decode/step3p5.py:_sample_draft_tokens_for_step` (Step3.5 MTP)

**The redistribution `[N, V/P] → [N/P, V]`** (all tokens × partial vocab → partial tokens × full vocab) is done with a `view + all_to_all_single + permute + view` sequence:
```
[N, V/P] -> view(P, N/P, V/P) -> all_to_all_single -> permute(1,0,2) -> view(N/P, V)
```
This is mathematically equivalent to the removed patch's `tensor_split(dim=0) + all_to_all(list) + cat(dim=-1)`, but keeps a single contiguous buffer for better HCCL fusion. Microbenchmarks on Atlas A3 (P=4, N=8–2048 tokens/rank) show **1.5x–2.4x speedup** over the list version, with the small-N shapes typical of decode gaining the most (2x+).

**Why not a bare `all_to_all_single`?** A naive `all_to_all_single` only redistributes on dim 0 (token) and never gathers the vocab dimension — it silently produces `[N, V/P]` instead of `[N/P, V]`. The `view` reshapes the input so dim 0 carries the per-rank token shard, letting the single collective route correctly; `permute` then interleaves the vocab shards of each token.

**Safety:** equal split is guaranteed because `pad_vocab_size` + `divide` make the vocab shard `V/P` identical on every rank. `world_size == 1` returns the input unchanged. For the dynamic token axis, `lmhead_all_to_all` raises a clear `ValueError` when `N` is not divisible by `world_size`; on the live lm-head path this is structural — `N` comes from `all_gather(hidden_states, dim=0)`, so `N = world_size × N_local` by construction. The three spec-decode call sites are defensive only: `enable_reduce_sample` and lm-head TP are mutually exclusive at startup (`ascend_config.py` raises `ValueError`), so those branches are unreachable today and need no round-up padding.

**Alignment with project direction:** the patch registration comment in `patch/__init__.py` already notes "Remove this patch when the refactor of all2all manager is done." This PR takes a step in that direction by removing the upper `GroupCoordinator` patch and using `dist.all_to_all_single` directly — the same pattern already used in `dsa_v1.py`, `dsa_cp.py`, `sfa_cp.py`, etc.

Also removes the patch registration comment block in `patch/__init__.py` and cleans up 4 affected test files.

### Does this PR introduce _any_ user-facing change?

No. The behavior of lm-head tensor parallel is unchanged; the refactor only replaces the internal communication call with an equivalent (and faster) one. No API, config, or model interface changes.

### How was this patch tested?

- **Lint:** `ruff check vllm_ascend/` clean, `ruff format --check` clean on all modified files.
- **Correctness simulation (4-rank, CPU):** verified the view-based helper is bit-identical to the removed list-based `tensor_split + all_to_all(list) + cat` implementation for P=4 / N=16 / V=8.
- **Unit tests (Ascend NPU, CI):** the `a2-1` selected-tests group covers the modified UT files (`test_vocab_parallel_embedding.py`, `test_patch_distributed.py`, `test_lmhead_all_to_all.py`) plus the Step3.5 MTP e2e.
- **Correctness microbench (2-rank gloo + 2/4-rank HCCL):** verified element-wise identical output to the old list-based version.
- **Performance microbench (Atlas A3, 4-rank HCCL, same-process A/B):** both implementations extracted from the respective source trees, outputs verified `allclose` equivalent; the single-buffer version is faster at every size:

  | N (tokens/rank) | list version | all_to_all_single | speedup |
  |---|---|---|---|
  | 8 | 0.311 ms | 0.131 ms | 2.36x |
  | 32 | 0.341 ms | 0.167 ms | 2.04x |
  | 128 | 0.254 ms | 0.161 ms | 1.58x |
  | 512 | 0.447 ms | 0.305 ms | 1.46x |
  | 2048 | 1.817 ms | 1.088 ms | 1.67x |
- **E2E, graph mode (Atlas A3, Qwen3-30B-A3B-W8A8, TP1×DP8 + lmhead_tp=4):** conversation checks (Chinese / English / math / code generation) all coherent, Qwen3 thinking behavior normal. A/B benchmark on identical config (random dataset, in=1024 / out=256, 64 prompts, 5 rounds each; the only difference is the vllm-ascend commit — PR `2c888b4d2` vs main @ merge-base `e80067d46`):

  | Metric | PR | baseline | Δ (5 rounds / warm-state) |
  |---|---|---|---|
  | TPOT | 44.30 ms | 45.91 ms | **-3.5% / -2.2%** |
  | Request throughput | 4.95 req/s | 4.80 req/s | **+3.1% / +2.4%** |
  | Output throughput | 1267.5 tok/s | 1228.9 tok/s | **+3.1% / +2.4%** |
  | Mean TTFT | 1381.6 ms | 1381.4 ms | flat (prefill untouched) |

  (warm-state = rounds 2–5, excluding the cold-start round)

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/58d3918e3ea0a544ffedadad2ba84559e9c51d8f

### CODE REVIEW

**Comment by xuchi-0808** [comment]
Addressed the divisibility check suggestion in commit 7a3bce88. Two adjustments from the original suggestion:

1. Used `raise ValueError(...)` instead of `assert` — matches the style used elsewhere in the same file (e.g. `_forward_embed_tp` after #10586) and survives `python -O`.
2. Moved the check before `vocab_per_partition = logits.shape[-1]` so the guard fires before any other shape-derived computation.

Also tightened the docstring: the previous wording claimed equal split was "guaranteed" on the token axis, which is not actually true at compile time — `N` is a dynamic request-batch size, so the divisibility is a runtime precondition, not a build-time invariant. The new docstring separates the two axes (vocab shard: guaranteed by `pad_vocab_size`; token count: runtime-checked) and explicitly notes why the guard exists (otherwise `view()` raises a cryptic shape error).

Thanks for catching this — without the guard, a future change that routes a non-divisible `N` through this path would fail with a confusing error.

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 1]
```diff
@@ -1174,7 +1175,7 @@ def _run_merged_draft(
             else:
                 logits = self.model.compute_logits(sample_hidden_states)
                 if lmhead_tp_enable():
-                    logits = get_lmhead_tp_group().all_to_all(logits)
+                    logits = lmhead_all_to_all(logits, get_lmhead_tp_group())
```
[the remark this answers is not in the document]
Comment (xuchi-0808) [reply vllm_ascend/spec_decode/llm_base_proposer.py:1153]: Good catch — adopted in commit 2411e5a3. Applied the round-up at both `max_num_reqs_across_dp` computation sites (lines 1085-1090 and 1276-1281), using `tp_size = get_lmhead_tp_group().world_size`.

Two notes on the analysis above, for the record:

1. The old list-based `tensor_split + all_to_all(list)` implementation technically did not crash on `N % P != 0` (because `torch.tensor_split` allows unequal chunks), but the resulting per-rank row counts would silently diverge — the downstream `_align_tensor_and_indices` assumes uniform shapes. So the old path was already broken in this corner case, just silently. The new `view + all_to_all_single` path fails fast instead, which is what surfaced this issue.

2. The `_get_logits_lmheadtp` site (`vocab_parallel_embedding.py`) does not need the same padding: its input always has shape `[N_local * P, V/P]` because it is computed from `all_gather(hidden_states, dim=0)` over the lmhead_tp group, so `N` is always a multiple of `P` by construction. Only the spec_decode path computes logits from a per-rank `sample_hidden_states` indexed by the padded `token_indices_to_sample`, so only it is affected.

The explicit `raise ValueError` divisibility check inside `lmhead_all_to_all` (from the previous review comment) is intentionally kept as defense-in-depth — if a future caller forgets to pad, we still fail loudly instead of producing silent misalignment.

**Comment by xuchi-0808** [comment]
Rebased onto the latest main (`7514442c5`, 2026-08-10). Changes in this round:

- Resolved the `npu_communicator.py` conflict by keeping the new `use_all2all` version gate while still removing the `all_to_all` method.
- Found and fixed a 4th consumer: `spec_decode/step3p5.py` (Step3.5 MTP, added by #10697) was still calling `GroupCoordinator.all_to_all`, which would break after removing the patch. Switched it to `lmhead_all_to_all` and added the same `max_num_reqs_across_dp` round-up fix.
- `pre-commit` and `main` checks are green; `select-tests` selected a test plan.

Could a maintainer add the `ready` label so the selected-tests jobs can run? Thanks!

**Review by wangxiyuan — APPROVED** [review]

**Comment by xuchi-0808** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: some jobs could not be re-run (still waiting on dependencies)

**Comment by xuchi-0808** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: some jobs could not be re-run (still waiting on dependencies)

**Comment by xuchi-0808** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: some jobs could not be re-run (still waiting on dependencies)
